File: services/python-validator/validator.py
def validate_granulometria(data):
    masses = data["retained_masses"]
    total_init_sample = data["total_sample_mass"]
    pan_mass = masses[-1]

    total_sample_real = sum(masses)
    fines_percent = (pan_mass / total_sample_real) * 100
    error_percentage = abs(total_sample_real - total_init_sample) / total_init_sample * 100

    valid_mass_balance = error_percentage <= 2

    if fines_percent <= 10:
        method_status = "Tamizado adecuado"
    else:
        method_status = "Advertencia: >10% de finos, se recomienda sedimentación complementaria"

    if valid_mass_balance:
        test_status = f"Práctica válida: el error de masa es {error_percentage:.2f}%, dentro de la tolerancia del 2%."
    else:
        test_status = f"Práctica inválida: el error de masa es {error_percentage:.2f}%, supera la tolerancia del 2%."

    retained_percentage_raw = [(mass / total_sample_real) * 100 for mass in masses]

    cumulative_raw = []
    running = 0
    for p in retained_percentage_raw:
        running += p
        cumulative_raw.append(min(running, 100))

    passing_raw = [max(0, 100 - c) for c in cumulative_raw]

    score = 1 if valid_mass_balance else 0

    return {
        "initial_sample": total_init_sample,
        "total_sample_real": total_sample_real,
        "p_error": round(error_percentage, 2),
        "retained_percentage": [round(x, 2) for x in retained_percentage_raw],
        "cumulative_percentage": [round(x, 2) for x in cumulative_raw],
        "passing_percentages": [round(x, 2) for x in passing_raw],
        "fines_percent": round(fines_percent, 2),
        "method_status": method_status,
        "valid_mass_balance": valid_mass_balance,
        "test_status": test_status,
        "score": score
    }

File: lib/validator.ts
import type { PracticeData, ValidationResult, CostEstimate } from "./types"

// Client-side validation logic (mirrors Python backend)
export function validateGranulometria(data: PracticeData): ValidationResult {
  const masses = data.retainedMasses
  const totalInitSample = data.totalSampleMass
  const panMass = masses[masses.length - 1]

  const totalSampleReal = masses.reduce((a, b) => a + b, 0)
  const finesPercent = (panMass / totalSampleReal) * 100
  const errorPercentage = Math.abs(totalSampleReal - totalInitSample) / totalInitSample * 100

  const validMassBalance = errorPercentage <= 2

  const methodStatus = finesPercent <= 10
    ? "Tamizado adecuado"
    : "Advertencia: >10% de finos, se recomienda sedimentacion complementaria"

  const testStatus = validMassBalance
    ? `Practica valida: error de masa ${errorPercentage.toFixed(2)}%, dentro de tolerancia del 2%.`
    : `Practica invalida: error de masa ${errorPercentage.toFixed(2)}%, supera tolerancia del 2%.`

  const retainedPercentage = masses.map(mass => (mass / totalSampleReal) * 100)

  const cumulativePercentage: number[] = []
  let running = 0
  for (const p of retainedPercentage) {
    running += p
    cumulativePercentage.push(Math.min(running, 100))
  }

  const passingPercentages = cumulativePercentage.map(c => Math.max(0, 100 - c))

  return {
    initialSample: totalInitSample,
    totalSampleReal,
    pError: Number(errorPercentage.toFixed(2)),
    retainedPercentage: retainedPercentage.map(x => Number(x.toFixed(2))),
    cumulativePercentage: cumulativePercentage.map(x => Number(x.toFixed(2))),
    passingPercentages: passingPercentages.map(x => Number(x.toFixed(2))),
    finesPercent: Number(finesPercent.toFixed(2)),
    methodStatus,
    validMassBalance,
    testStatus,
    score: validMassBalance ? 1 : 0,
  }
}

// Cost estimation for Solana transactions
export function estimateCosts(solPrice: number = 150): CostEstimate {
  const transactionFee = 0.000005 // ~5000 lamports base fee
  const rentExempt = 0.00089 // ~890,880 lamports for small account
  const programExecution = 0.00002 // Estimated compute units

  const total = transactionFee + rentExempt + programExecution
  const totalUsd = total * solPrice

  return {
    transactionFee,
    rentExempt,
    programExecution,
    total,
    totalUsd,
  }
}
